File: backend/src/morado/api/v1/report.py
"""Test Report API endpoints.

This module provides REST API endpoints for generating test reports and analytics.
"""

from datetime import datetime
from typing import Annotated, Any

from litestar import Controller, get
from litestar.di import Provide
from litestar.params import Parameter
from morado.services.report import ReportService
from sqlalchemy.orm import Session


def provide_report_service() -> ReportService:
    """Provide ReportService instance."""
    return ReportService()


class ReportController(Controller):
    """Controller for Report generation endpoints."""

    path = "/reports"
    tags = ["Reports"]
    dependencies = {"report_service": Provide(provide_report_service)}

    @get("/execution-summary")
    async def get_execution_summary_report(
        self,
        report_service: ReportService,
        db_session: Session,
        start_date: Annotated[datetime | None, Parameter(query="start_date")] = None,
        end_date: Annotated[datetime | None, Parameter(query="end_date")] = None,
        environment: Annotated[str | None, Parameter(query="environment")] = None,
        test_case_id: Annotated[int | None, Parameter(query="test_case_id")] = None,
        test_suite_id: Annotated[int | None, Parameter(query="test_suite_id")] = None,
    ) -> dict[str, Any]:
        """Get execution summary report.

        This endpoint provides aggregated statistics about test executions
        within a specified time period.

        Args:
            report_service: Report service instance
            db_session: Database session
            start_date: Start date for filtering
            end_date: End date for filtering
            environment: Filter by environment
            test_case_id: Filter by test case ID
            test_suite_id: Filter by test suite ID

        Returns:
            Execution summary statistics

        Example:
            GET /reports/execution-summary?start_date=2024-01-01&end_date=2024-01-31
        """
        return report_service.get_execution_summary_report(
            db_session,
            start_date=start_date,
            end_date=end_date,
            environment=environment,
            test_case_id=test_case_id,
            test_suite_id=test_suite_id
        )

    @get("/test-case/{test_case_id:int}")
    async def get_test_case_report(
        self,
        test_case_id: int,
        report_service: ReportService,
        db_session: Session,
        limit: Annotated[int, Parameter(query="limit", ge=1, le=100)] = 10,
    ) -> dict[str, Any]:
        """Get test case execution report.

        This endpoint provides execution history and statistics for a specific test case.

        Args:
            test_case_id: Test case ID
            report_service: Report service instance
            db_session: Database session
            limit: Number of recent executions to include

        Returns:
            Test case execution history and statistics
        """
        return report_service.get_test_case_report(
            db_session,
            test_case_id,
            limit=limit
        )

    @get("/test-suite/{test_suite_id:int}")
    async def get_test_suite_report(
        self,
        test_suite_id: int,
        report_service: ReportService,
        db_session: Session,
        limit: Annotated[int, Parameter(query="limit", ge=1, le=100)] = 10,
    ) -> dict[str, Any]:
        """Get test suite execution report.

        This endpoint provides execution history and statistics for a specific test suite.

        Args:
            test_suite_id: Test suite ID
            report_service: Report service instance
            db_session: Database session
            limit: Number of recent executions to include

        Returns:
            Test suite execution history and statistics
        """
        return report_service.get_test_suite_report(
            db_session,
            test_suite_id,
            limit=limit
        )

    @get("/trend")
    async def get_trend_report(
        self,
        report_service: ReportService,
        db_session: Session,
        days: Annotated[int, Parameter(query="days", ge=1, le=365)] = 30,
        environment: Annotated[str | None, Parameter(query="environment")] = None,
    ) -> dict[str, Any]:
        """Get execution trend report.

        This endpoint provides daily execution trends over a specified period.

        Args:
            report_service: Report service instance
            db_session: Database session
            days: Number of days to include in trend
            environment: Filter by environment

        Returns:
            Daily execution trends

        Example:
            GET /reports/trend?days=30&environment=test
        """
        return report_service.get_trend_report(
            db_session,
            days=days,
            environment=environment
        )

    @get("/environment-comparison")
    async def get_environment_comparison_report(
        self,
        report_service: ReportService,
        db_session: Session,
        start_date: Annotated[datetime | None, Parameter(query="start_date")] = None,
        end_date: Annotated[datetime | None, Parameter(query="end_date")] = None,
    ) -> dict[str, Any]:
        """Get environment comparison report.

        This endpoint provides statistics comparing test executions across
        different environments.

        Args:
            report_service: Report service instance
            db_session: Database session
            start_date: Start date for filtering
            end_date: End date for filtering

        Returns:
            Statistics by environment

        Example:
            GET /reports/environment-comparison?start_date=2024-01-01
        """
        return report_service.get_environment_comparison_report(
            db_session,
            start_date=start_date,
            end_date=end_date
        )

    @get("/failure-analysis")
    async def get_failure_analysis_report(
        self,
        report_service: ReportService,
        db_session: Session,
        start_date: Annotated[datetime | None, Parameter(query="start_date")] = None,
        end_date: Annotated[datetime | None, Parameter(query="end_date")] = None,
        limit: Annotated[int, Parameter(query="limit", ge=1, le=100)] = 10,
    ) -> dict[str, Any]:
        """Get failure analysis report.

        This endpoint provides analysis of failed test executions,
        including error messages and failure patterns.

        Args:
            report_service: Report service instance
            db_session: Database session
            start_date: Start date for filtering
            end_date: End date for filtering
            limit: Number of top failures to include

        Returns:
            Failure analysis

        Example:
            GET /reports/failure-analysis?limit=20
        """
        return report_service.get_failure_analysis_report(
            db_session,
            start_date=start_date,
            end_date=end_date,
            limit=limit
        )

File: frontend/src/api/report.ts
/**
 * Report API Client
 *
 * Test Report and Analytics
 * Provides functions for generating test reports and analytics.
 */

import apiClient from './index';

// Report Types
export interface ExecutionSummaryReport {
  total_executions: number;
  passed_executions: number;
  failed_executions: number;
  success_rate: number;
  average_duration: number;
  total_duration: number;
  environment?: string;
  start_date?: string;
  end_date?: string;
  test_case_count?: number;
  test_suite_count?: number;
}

export interface TestCaseReport {
  test_case_id: number;
  test_case_name: string;
  total_executions: number;
  passed_executions: number;
  failed_executions: number;
  success_rate: number;
  average_duration: number;
  last_execution?: {
    timestamp: string;
    success: boolean;
    duration: number;
    error?: string;
  };
  recent_executions: Array<{
    execution_id: number;
    timestamp: string;
    success: boolean;
    duration: number;
    environment: string;
  }>;
}

export interface TestSuiteReport {
  test_suite_id: number;
  test_suite_name: string;
  total_executions: number;
  passed_executions: number;
  failed_executions: number;
  success_rate: number;
  average_duration: number;
  test_case_count: number;
  last_execution?: {
    timestamp: string;
    success: boolean;
    duration: number;
    passed_count: number;
    failed_count: number;
  };
  recent_executions: Array<{
    execution_id: number;
    timestamp: string;
    success: boolean;
    duration: number;
    environment: string;
    passed_count: number;
    failed_count: number;
  }>;
}

export interface TrendReport {
  days: number;
  environment?: string;
  daily_trends: Array<{
    date: string;
    total_executions: number;
    passed_executions: number;
    failed_executions: number;
    success_rate: number;
    average_duration: number;
  }>;
}

export interface EnvironmentComparisonReport {
  start_date?: string;
  end_date?: string;
  environments: Array<{
    environment: string;
    total_executions: number;
    passed_executions: number;
    failed_executions: number;
    success_rate: number;
    average_duration: number;
  }>;
}

export interface FailureAnalysisReport {
  start_date?: string;
  end_date?: string;
  total_failures: number;
  top_failures: Array<{
    test_case_id: number;
    test_case_name: string;
    failure_count: number;
    failure_rate: number;
    common_errors: Array<{
      error_message: string;
      count: number;
    }>;
    last_failure: {
      timestamp: string;
      error: string;
      environment: string;
    };
  }>;
}

export interface ExecutionSummaryParams {
  start_date?: string;
  end_date?: string;
  environment?: string;
  test_case_id?: number;
  test_suite_id?: number;
}

export interface TrendReportParams {
  days?: number;
  environment?: string;
}

export interface EnvironmentComparisonParams {
  start_date?: string;
  end_date?: string;
}

export interface FailureAnalysisParams {
  start_date?: string;
  end_date?: string;
  limit?: number;
}

// Report API Functions

/**
 * Get execution summary report
 * Provides aggregated statistics about test executions within a specified time period
 */
export async function getExecutionSummaryReport(
  params?: ExecutionSummaryParams,
): Promise<ExecutionSummaryReport> {
  const response = await apiClient.get<ExecutionSummaryReport>('/reports/execution-summary', {
    params,
  });
  return response.data;
}

/**
 * Get test case execution report
 * Provides execution history and statistics for a specific test case
 */
export async function getTestCaseReport(testCaseId: number, limit = 10): Promise<TestCaseReport> {
  const response = await apiClient.get<TestCaseReport>(`/reports/test-case/${testCaseId}`, {
    params: { limit },
  });
  return response.data;
}

/**
 * Get test suite execution report
 * Provides execution history and statistics for a specific test suite
 */
export async function getTestSuiteReport(
  testSuiteId: number,
  limit = 10,
): Promise<TestSuiteReport> {
  const response = await apiClient.get<TestSuiteReport>(`/reports/test-suite/${testSuiteId}`, {
    params: { limit },
  });
  return response.data;
}

/**
 * Get execution trend report
 * Provides daily execution trends over a specified period
 */
export async function getTrendReport(params?: TrendReportParams): Promise<TrendReport> {
  const response = await apiClient.get<TrendReport>('/reports/trend', { params });
  return response.data;
}

/**
 * Get environment comparison report
 * Provides statistics comparing test executions across different environments
 */
export async function getEnvironmentComparisonReport(
  params?: EnvironmentComparisonParams,
): Promise<EnvironmentComparisonReport> {
  const response = await apiClient.get<EnvironmentComparisonReport>(
    '/reports/environment-comparison',
    { params },
  );
  return response.data;
}

/**
 * Get failure analysis report
 * Provides analysis of failed test executions, including error messages and failure patterns
 */
export async function getFailureAnalysisReport(
  params?: FailureAnalysisParams,
): Promise<FailureAnalysisReport> {
  const response = await apiClient.get<FailureAnalysisReport>('/reports/failure-analysis', {
    params,
  });
  return response.data;
}

/**
 * Export report to CSV
 */
export async function exportReportToCSV(
  reportType: 'execution-summary' | 'trend' | 'environment-comparison' | 'failure-analysis',
  params?: Record<string, unknown>,
): Promise<Blob> {
  const response = await apiClient.get(`/reports/${reportType}/export`, {
    params,
    responseType: 'blob',
  });
  return response.data;
}

/**
 * Export report to PDF
 */
export async function exportReportToPDF(
  reportType: 'execution-summary' | 'trend' | 'environment-comparison' | 'failure-analysis',
  params?: Record<string, unknown>,
): Promise<Blob> {
  const response = await apiClient.get(`/reports/${reportType}/export-pdf`, {
    params,
    responseType: 'blob',
  });
  return response.data;
}

/**
 * Get dashboard summary
 * Provides a quick overview of key metrics for the dashboard
 */
export async function getDashboardSummary(environment?: string): Promise<{
  total_test_cases: number;
  total_test_suites: number;
  recent_executions: number;
  success_rate: number;
  average_duration: number;
  trending_failures: Array<{
    test_case_id: number;
    test_case_name: string;
    failure_count: number;
  }>;
}> {
  const response = await apiClient.get('/reports/dashboard-summary', {
    params: { environment },
  });
  return response.data;
}
